src/clib/pioc_support.c

```diff
@@ -4046,24 +4048,36 @@ int pioc_change_def(int ncid, int is_enddef)
          * space in the header when creating NetCDF files. The current calls
          * to nc_enddef()/ncmpi_enddef() need to be replaced with nc__enddef
          * ()/ncmpi__enddef() (note the double underscore).
+         *
+         * This reservation should be made only once: after the header section
+         * is expanded, a new reservation may involve moving (shifting) data,
+         * which can be very expensive if the data sections are huge.
          */
 #ifdef _PNETCDF
         if (file->iotype == PIO_IOTYPE_PNETCDF)
         {
             if (is_enddef)
             {
-                /* Sets the pad at the end of the "header" section. */
-                const MPI_Offset h_minfree = 10 * 1024; /* The recommended size by Charlie Zender (NCO developer) is 10 KB */
+                if (file->reserve_extra_header_space)
+                {
+                    /* Sets the pad at the end of the "header" section.
+                     * The recommended size by Charlie Zender (NCO developer) is 10 KB */
+                    const MPI_Offset h_minfree = (PIO_RESERVED_FILE_HEADER_SIZE >= 0)? PIO_RESERVED_FILE_HEADER_SIZE : (10 * 1024);
```

Comment:
You can move sanity check of PIO_RESERVED_FILE_HEADER_SIZE to configure (if its < 0 set it to default in configure and warn the user - the warning is important) and just add an assert here.

Also PIO_RESERVED_FILE_HEADER_SIZE == 0 should be a valid option (the user wants to disable allocating any extra space in the header)

Reply:
OK, I have made this change.